### PR TITLE
Add pre-commit hook for TheAlgorithms/Python

### PR DESCRIPTION
**Pull request by dhruvmanila**
As mentioned in https://github.com/TheAlgorithms/Python/pull/2509#issuecomment-700778237

### **Describe your change:**

* [x] Add an algorithm?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by dhruvmanila** [comment]
There are a bunch of files that will be fixed if I run this.
Should I open another PR for those fixes?

**Comment by dhruvmanila** [comment]
We also need to inform people to enable it in their local copy by:
```console
$ pip install pre-commit
$ pre-commit install
```

Also, there are badges for [pre-commit](https://pre-commit.com/#badging-your-repository) and [black](https://img.shields.io/badge/code%20style-black-000000.svg).
Shall I add them?

File: .pre-commit-config.yaml
Comment (cclauss) [inline .pre-commit-config.yaml:5]: Do we have something in CONTRIBUTING.md that describes the proper shebang
(`#!/usr/bin/env python3`)?

**Review by cclauss — CHANGES_REQUESTED** [review]
This is awesome!  How long does it take to run on your machine?  (If it is not fast, contributors will turn it off.)  Please add the badges to README.md in a separate follow-up PR.  Please fix the issues in a separate PR that we should land before this one.

File: .pre-commit-config.yaml
Comment (dhruvmanila) [reply .pre-commit-config.yaml:5]: No. I thought as executable files are made by us for mostly testing purposes, to include this in case we forget to add it.
Should I remove it?

**Comment by dhruvmanila** [comment]
By default, `pre-commit` runs only on the new files or modified files. We can change it to run on all the files all the time but that would be inefficient.

File: .pre-commit-config.yaml
Comment (cclauss) [reply .pre-commit-config.yaml:5]: How many files will this flag?

**Comment by cclauss** [comment]
OK.  Yes now I understand.  After this lands, we can add something like this to do all files.  https://github.com/cclauss/itinerant-tester/blob/master/.github/workflows/pre-commit.yml

File: .pre-commit-config.yaml
Comment (dhruvmanila) [reply .pre-commit-config.yaml:5]: 3 files:
- machine_learning/scoring_functions.py
- maths/sum_of_arithmetic_series.py
- scheduling/round_robin.py

Comment (cclauss) [reply .pre-commit-config.yaml:5]: Let’s leave it in.

**Comment by dhruvmanila** [comment]
I don't think so we need a workflow file for this as all this is done locally. The workflow file is run after the commit has been pushed and mainly used to check if every test passed. It's mostly just for the visuals as far as I can understand.

**Comment by dhruvmanila** [comment]
TODO:
[ ] Separate PR for all the fixes BEFORE this is merged
[ ] Follow-up PR for adding badges to README.md

Is that correct?

**Comment by cclauss** [comment]
A fraction of all contributors will install and run pre-commit.  The workflow would check the work of the contributors wo do not install the tool and it will allow us to remove all that complexity from .travis.yml which should speed up our test runs.

**Comment by dhruvmanila** [comment]
Got it!

**Review by cclauss — APPROVED** [review]

**Comment by cclauss** [comment]
Do you have rights to click the `Squash and merge` button above?

**Comment by dhruvmanila** [comment]
Missed it! But I don't think so as that will be available only to members and owners.

**Comment by cclauss** [comment]
OK.  We need the fix PR before we merge.

Also I added to the GitHub Action to your task list above.

File: .pre-commit-config.yaml
Comment (dhruvmanila) [reply .pre-commit-config.yaml:5]: So, shall I include the shebang in those files or remove the executable bit from them?

**Comment by dhruvmanila** [comment]
@cclauss DON'T MERGE THIS YET.
We need to ignore some files which require whitespace and newlines.

**Comment by dhruvmanila** [comment]
@cclauss Please take a look at the changes:
EOF newline will be checked only for Python files and ignoring whitespace for that specific file as it requires it in `pytest`

**Comment by dhruvmanila** [comment]
I can see the merge option on the app but not on the browser.

**Comment by cclauss** [comment]
THANKS!!

**Comment by DmytroLitvinov** [comment]
Hi @dhruvmanila ,

I guess we should warn peope about pre-commit at [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md) file.

**Comment by cclauss** [comment]
Let's not force the complexities of a local install of pre-commit on all contributors.  Some folks are pretty junior and adding pre-commit to their machines might turn them off from contributing.  The README should however show them how to find and solve issues when the GitHub Action does not love their contribution.

**Comment by dhruvmanila** [comment]
Maybe we should put it in `Travis_CI_tests_are_failing` and change the name of the file to `tests_are_failing` and put the instructions as to how to find the error for all the automatic tests we're performing in that file.

The other idea would be to separate out the files for each test but having a lot of files doesn't seem attractive.

**Pull request merged**
